### PR TITLE
fix: incorrect output scaling when input is less than desired amount

### PR DESCRIPTION
**Pull request by kimpers**
## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts [hunk 1]
```diff
@@ -275,16 +275,14 @@ function findRoutesAndCreateOptimalPath(
         }
 
         // TODO(kimpers): remove once we have solved the rounding/precision loss issues in the Rust router
-        const scaleOutput = (fillInput: BigNumber, output: BigNumber) =>
-            output
-                .dividedBy(fillInput)
-                .times(rustInputAdjusted)
-                .decimalPlaces(0, side === MarketOperation.Sell ? BigNumber.ROUND_FLOOR : BigNumber.ROUND_CEIL);
+        const maxSampledOutput = BigNumber.max(...routeSamples.map(s => s.output));
+        const scaleOutput = (output: BigNumber) => BigNumber.min(output.times(scale), maxSampledOutput);
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts:279]: Wouldn't we want to min against the sum of outputs? We can still use multiple sources/samples even if we don't hit our Target, right?

**Review by dorothy-zbornak — COMMENTED** [review]
When does this come up? Don't we abort with insufficient liquidity if we can't satisfy the input amount?

File: packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts (hunk 1 above)
Comment (kimpers) [reply packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts:279]: This is all a mess and if it weren't for the f64 precision issues we wouldn't need this duct tape.

I could be wrong but this scaling is on a per fill level, and therefore it should always be from a single source, within the sampled range of that source, correct?

I added this min to check that we don't scale the output outside of the sampled range, where the liquidity is undefined.

**Comment by kimpers** [comment]
> When does this come up? Don't we abort with insufficient liquidity if we can't satisfy the input amount?

This was observed on Optimism, due to sampling running out of gas because it had to cross too many ticks on Uni V3 (only source at this time).

As it stands today we don't abort with insufficient liquidity in this scenario (but maybe we should). One of the assumptions in the router is that e.g for sells you can always sell more for the same output. So if we have the last sample that sells 10k DAI -> 2 ETH, it assumes that it can also sell 20k DAI -> 2 ETH and will therefore try to Route in the best possible way with the samples it received.

**Review by dorothy-zbornak — APPROVED** [review]

**Pull request merged**